### PR TITLE
Update Dagger Injection Process

### PR DESCRIPTION
**Pull request by leo0263**
- Inject ViewModelFactory on MainActivity
- Inject GetComicByNumber and GetLatestComic on MainViewModel\
- Inject Schedulers.Io
- Inject AndroidSchedulers.mainThread()
- Create RetrofitModule that provide the xkcdService
- Inject xkcdService on MainViewModelModule via including RetrofitModule

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -70,9 +70,13 @@ dependencies {
     // add mockito kotlin
     testImplementation "com.nhaarman:mockito-kotlin:1.6.0"
 
-    // gradle 2
-    implementation 'com.google.dagger:dagger:2.19'
-    kapt 'com.google.dagger:dagger-compiler:2.19'
+    // Dagger -> kept at 2.16 because https://github.com/google/dagger/issues/1245
```
Comment (bopbi) [inline app/build.gradle]: :+1:

File: app/src/main/java/com/arjunalabs/android/xkcdreader/MyApplication.kt [hunk 2]
```diff
@@ -1,24 +1,28 @@
 package com.arjunalabs.android.xkcdreader
 
+import android.app.Activity
```
Comment (bopbi) [inline app/src/main/java/com/arjunalabs/android/xkcdreader/MyApplication.kt]: :+1:

File: app/src/main/java/com/arjunalabs/android/xkcdreader/dagger/MainViewModelModule.kt [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    @Provides
+    @Named("schedulersIo")
+    fun providesSchedulersIo(): Scheduler = Schedulers.io()
+
+    @Provides
+    @Named("schedulersMainThread")
+    fun providesSchedulersMainThread(): Scheduler = AndroidSchedulers.mainThread()
+}
```
Comment (bopbi) [inline app/src/main/java/com/arjunalabs/android/xkcdreader/dagger/MainViewModelModule.kt]: :+1:

File: app/src/main/java/com/arjunalabs/android/xkcdreader/dagger/MyApplicationComponent.kt [hunk 4]
```diff
@@ -1,13 +1,25 @@
...
 
-interface MyApplicationComponent {
+interface MyApplicationComponent : AndroidInjector<MyApplication> {
 
-    fun inject(mainActivity: MainActivity)
+    @Component.Builder
+    abstract class Builder : AndroidInjector.Builder<MyApplication>() {
+        @BindsInstance
+        abstract fun appContext(appContext: Context): Builder
+    }
```
Comment (bopbi) [inline app/src/main/java/com/arjunalabs/android/xkcdreader/dagger/MyApplicationComponent.kt]: :+1:

File: app/src/main/java/com/arjunalabs/android/xkcdreader/dagger/MyApplicationModule.kt [hunk 5]
```diff
@@ -1,17 +1,12 @@
...
-import com.arjunalabs.android.xkcdreader.repository.XKCDServiceImpl
+import com.arjunalabs.android.xkcdreader.ui.MainActivity
 import dagger.Module
-import dagger.Provides
-import javax.inject.Singleton
+import dagger.android.ContributesAndroidInjector
 
 @Module
-open class MyApplicationModule {
+abstract class MyApplicationModule {
```
Comment (bopbi) [inline app/src/main/java/com/arjunalabs/android/xkcdreader/dagger/MyApplicationModule.kt]: :+1:

File: app/src/main/java/com/arjunalabs/android/xkcdreader/dagger/RetrofitModule.kt [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+import dagger.Provides
+
+@Module
+class RetrofitModule {
+
+    @Provides
+    open fun provideXKCDService(): XKCDService {
+        return XKCDServiceImpl.create()
+    }
+}
```
Comment (bopbi) [inline app/src/main/java/com/arjunalabs/android/xkcdreader/dagger/RetrofitModule.kt]: :+1:

File: app/src/main/java/com/arjunalabs/android/xkcdreader/dagger/ViewModelFactory.kt [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+class ViewModelFactory<VM : ViewModel> @Inject constructor(
+        private val viewModel: Lazy<VM>
+) : ViewModelProvider.Factory {
+
+    @Suppress("UNCHECKED_CAST")
+    override fun <T : ViewModel> create(modelClass: Class<T>): T {
+        return viewModel.get() as T
+    }
+}
```
Comment (bopbi) [inline app/src/main/java/com/arjunalabs/android/xkcdreader/dagger/ViewModelFactory.kt]: :+1:

File: app/src/main/java/com/arjunalabs/android/xkcdreader/ui/MainActivity.kt [hunk 8]
```diff
@@ -71,7 +67,7 @@ class MainActivity : AppCompatActivity() {
         mainActivityState.let {
 
             when (it) {
-                is  MainActivityState.Error -> {
+                is MainActivityState.Error -> {
```
Comment (bopbi) [inline app/src/main/java/com/arjunalabs/android/xkcdreader/ui/MainActivity.kt]: :+1:

File: app/src/main/java/com/arjunalabs/android/xkcdreader/ui/MainViewModel.kt [hunk 9]
```diff
@@ -8,17 +8,17 @@ import com.arjunalabs.android.xkcdreader.usecase.GetLatestComic
...
+import javax.inject.Named
 
-class MainViewModel(
+class MainViewModel @Inject constructor(
         private val getComicByNumber: GetComicByNumber,
         private val getLatestComic: GetLatestComic,
-        private val subscriberSchedulers: Scheduler = Schedulers.io(),
-        private val observerSchedulers: Scheduler = AndroidSchedulers.mainThread()) : ViewModel(),
-        StateObservable<MainActivityState> {
+        @Named("schedulersIo") private val subscriberSchedulers: Scheduler,
```
Comment (bopbi) [inline app/src/main/java/com/arjunalabs/android/xkcdreader/ui/MainViewModel.kt]: :+1: i never think about this one

**Review by bopbi — APPROVED** [review]
Great

File: app/src/main/java/com/arjunalabs/android/xkcdreader/ui/MainViewModel.kt (hunk 9 above)
Comment (leo0263) [reply app/src/main/java/com/arjunalabs/android/xkcdreader/ui/MainViewModel.kt:19]: TIL 🙇

**Pull request merged**